lib/compiler-webpack5.js

```diff
@@ -0,0 +1,374 @@
+      // Output name
+      outputName: outputName,
+      // Compiled code
+      content: compiledTemplate.content
+    };
+  });
+}
+
+/**
+ * Return all file dependencies of the last child compilation
+ *
+ * @param {WebpackCompiler} compiler
+ * @returns {Array<string>}
+ */
+function getFileDependencies (compiler) {
+  const childCompiler = getChildCompiler(compiler);
+  return childCompiler.fileDependencies;
+}
+
+/**
+ * @type {WeakMap<object, Snapshot>}
+ */
+const snapshotMap = new WeakMap();
+/**
+ * Calls the callback with `true` if the file dependencies of the current childCompiler
+ * for the given mainCompilation are valid.
+ *
+ * Uses the `snapshotMap` cache if possible.
+ *
+ * @param {WebpackCompilation} mainCompilation
+ * @param callback
+ */
+function hasOutDatedTemplateCache (mainCompilation, callback) {
+  const childCompiler = getChildCompiler(mainCompilation.compiler);
+
+  function updateSnapshot (snapshotValidErr, snapshotValid) {
+    return mainCompilation.fileSystemInfo.createSnapshot(
+      Date.now(),
+      childCompiler.fileDependencies,
+      childCompiler.contextDependencies,
+      childCompiler.missingDependencies,
+      null,
+      (err, snapshot) => {
+        // console.log(snapshot);
+        if (err) {
+          return callback(err);
+        }
+        snapshotMap.set(childCompiler, snapshot);
+        if (isChildCompilerCacheOutdated(childCompiler, snapshot)) {
+          return callback(null, false);
+        } else {
+          return callback(snapshotValidErr, snapshotValid);
+        }
+      }
+    );
+  }
+
+  // Try to get the `checkChildCompilerCache` result from cache
+  const prevSnapshot = snapshotMap.get(childCompiler);
+  if (prevSnapshot !== undefined) {
+    return mainCompilation.fileSystemInfo.checkSnapshotValid(
+      prevSnapshot,
+      (err, isValid) => {
+        return updateSnapshot(err, isValid);
+      }
+    );
+  }
+
+  return updateSnapshot(null, false);
+}
+
+/**
+ * Returns `true` if the file dependencies of the given childCompiler are outdated.
+ *
+ * @param {WebpackCompilation} mainCompilation
+ * @param {HtmlWebpackChildCompiler} childCompiler
+ * @returns {boolean}
+ */
+function isChildCompilerCacheOutdated (childCompiler, snapshot) {
+  // If the compilation was never run there is no invalid cache
+  if (!childCompiler.compilationStartedTimestamp) {
+    return false;
+  }
+  // Check if any dependent file was changed after the last compilation
+  const fileTimestamps = snapshot.fileTimestamps;
+  const isCacheOutOfDate = childCompiler.fileDependencies.some((fileDependency) => {
+    console.log(fileDependency);
+    const timestamp = fileTimestamps.get(fileDependency).timestamp;
+    // If the timestamp is not known the file is new
+    // If the timestamp is larger then the file has changed
+    // Otherwise the file is still the same
+    return !timestamp || timestamp > childCompiler.compilationStartedTimestamp;
+  });
+  return isCacheOutOfDate;
```

Comment:
Wow you really made it to be sync!

That's amazing :)

I'll have to look into this but this would solve it really nicely

Reply:
Hm, I'm not so sure, I think the compilation starts before this check is done.

timestamp > childCompiler.compilationStartedTimestamp is false :(